File: src/main/python/dfs/number_of_enclaves.py
from typing import *


class NumberOfEnclaves:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        def dfs(r, c):
            # if out of bounds or 0 return
            if r < 0 or c < 0 or r == len(grid) or c == len(grid) or grid[r][c] != 1:
                return
            # if 1, set it to 0
            grid[r][c] = 0
            # call dfs on all the neighbors
            dfs(r + 1, c), dfs(r - 1, c), dfs(r, c + 1), dfs(r, c - 1)

        # convert all the 1's that are connected to the boundary to 0's
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if i * j == 0 or i == len(grid) - 1 or j == len(grid[0]) - 1:
                    dfs(i, j)

        # count the entries in the grid that are 1s
        return sum([sum(row) for row in grid])


if __name__ == "__main__":
    s = NumberOfEnclaves()
    e1 = s.numEnclaves([[0, 0, 0, 0], [1, 0, 1, 0], [0, 1, 1, 0], [0, 0, 0, 0]])
    print(e1 == 3)
    e2 = s.numEnclaves([[0, 1, 1, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0, 0, 0, 0]])
    print(e2 == 0)

File: src/main/java/org/parthvnp/DFS/NumberOfEnclaves.java
package org.parthvnp.DFS;

import java.util.Arrays;

public class NumberOfEnclaves {
    public int numEnclaves(int[][] grid) {
        int count = 0;
        // convert all boundary 1s to 0s and all the 1s reachable from boundary to 0s
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (i * j == 0 || i == grid.length - 1 || j == grid[0].length - 1) dfs(grid, i, j);
            }
        }
        // count the remaining 1s
        return Arrays.stream(grid).flatMapToInt(Arrays::stream).sum();
    }

    public void dfs(int[][] grid, int row, int col) {
        // if the current is out of bounds of not 1 then return
        if (row < 0 || col < 0 || row == grid.length || col == grid[0].length || grid[row][col] == 0) return;
        // the current cell is within bounds and is 1 so change it to 0
        grid[row][col] = 0;
        // now traverse in all the directions
        dfs(grid, row + 1, col);
        dfs(grid, row - 1, col);
        dfs(grid, row, col + 1);
        dfs(grid, row, col - 1);
    }

    public static void main(String[] args) {
        var s = new NumberOfEnclaves();
        var e1 = s.numEnclaves(new int[][]{{0, 0, 0, 0}, {1, 0, 1, 0}, {0, 1, 1, 0}, {0, 0, 0, 0}});
        System.out.println(e1 == 3);
    }
}
